vms/platformvm/validators/manager.go

```diff
@@ -170,56 +171,105 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
 		return nil, ErrMissingValidator
 	}
 
-	currentSubnetValidatorList := currentSubnetValidators.List()
-	vdrSet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentSubnetValidatorList))
-	for _, vdr := range currentSubnetValidatorList {
-		primaryVdr, ok := currentPrimaryNetworkValidators.Get(vdr.NodeID)
+	currentSubnetValidatorList := currentSubnetValidatorSet.List()
+	currentSubnetValidators := make([]*maskedValidator, 0, len(currentSubnetValidatorList))
+	currentSubnetValidatorIndices := make(map[ids.NodeID]int)
+	for i, validator := range currentSubnetValidatorList {
+		primaryVdr, ok := currentPrimaryNetworkValidators.Get(validator.NodeID)
 		if !ok {
 			// This should never happen
-			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, vdr.NodeID)
-		}
-		vdrSet[vdr.NodeID] = &validators.GetValidatorOutput{
-			NodeID:    vdr.NodeID,
-			PublicKey: primaryVdr.PublicKey,
-			Weight:    vdr.Weight,
+			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, validator.NodeID)
 		}
+
+		currentSubnetValidators = append(currentSubnetValidators, &maskedValidator{
+			GetValidatorOutput: &validators.GetValidatorOutput{
+				NodeID:    validator.NodeID,
+				PublicKey: primaryVdr.PublicKey,
+				Weight:    validator.Weight,
+			},
+			masked: false,
+		})
+		currentSubnetValidatorIndices[validator.NodeID] = i
 	}
 
+	validatorsFromDiffs := make([]*maskedValidator, 0)
+	validatorsFromDiffsIndices := make(map[ids.NodeID]int)
 	for diffHeight := lastAcceptedHeight; diffHeight > height; diffHeight-- {
-		err := m.applyValidatorDiffs(vdrSet, subnetID, diffHeight)
-		if err != nil {
+		if validatorsFromDiffs, err = m.applyValidatorDiffs(
+			currentSubnetValidators,
+			currentSubnetValidatorIndices,
+			validatorsFromDiffs,
+			validatorsFromDiffsIndices,
+			subnetID,
+			diffHeight,
+		); err != nil {
 			return nil, err
 		}
 	}
 
+	// The current validator set is guaranteed to be sorted. Sort the set of
+	// validators calculated from applying diffs, so we can merge the two sorted
+	// lists.
+	sort.Slice(validatorsFromDiffs, func(i, j int) bool {
+		return validatorsFromDiffs[i].less(validatorsFromDiffs[j])
+	})
+
+	result := merge(currentSubnetValidators, validatorsFromDiffs)
+
 	// cache the validator set
-	validatorSetsCache.Put(height, vdrSet)
+	validatorSetsCache.Put(height, result)
 
 	endTime := m.clk.Time()
 	m.metrics.IncValidatorSetsCreated()
 	m.metrics.AddValidatorSetsDuration(endTime.Sub(startTime))
 	m.metrics.AddValidatorSetsHeightDiff(lastAcceptedHeight - height)
-	return vdrSet, nil
+	return result, nil
 }
 
 func (m *manager) applyValidatorDiffs(
-	vdrSet map[ids.NodeID]*validators.GetValidatorOutput,
+	currentSubnetValidators []*maskedValidator,
+	currentSubnetValidatorIndices map[ids.NodeID]int,
+	validatorsFromDiffs []*maskedValidator,
```

Comment:
could we add a comment that `validatorsFromDiffs` only contains validators that are not part of `currentSubnetValidators`?